New resource: azurerm_managed_application (#6386)
terraform-providers/terraform-provider-azurerm: website/docs/r/managed_application.html.markdown

  }
}

resource "azurerm_managed_application" "example" {
  name                        = "example-managedapplication"
  location                    = azurerm_resource_group.example.location
  resource_group_name         = azurerm_resource_group.example.name
  kind                        = "ServiceCatalog"
  managed_resource_group_name = "infrastructureGroup"
  application_definition_id   = azurerm_managed_application_definition.example.id

  parameters = {
    location                 = azurerm_resource_group.example.location
    storageAccountNamePrefix = "storeNamePrefix"
    storageAccountType       = "Standard_LRS"
  }
}
```

## Argument Reference

The following arguments are supported:

* `name` - (Required) Specifies the name of the Managed Application. Changing this forces a new resource to be created.

* `resource_group_name` - (Required) The name of the Resource Group where the Managed Application should exist. Changing this forces a new resource to be created.

* `location` - (Required) Specifies the supported Azure location where the resource exists. Changing this forces a new resource to be created.

> * `kind` - (Required) The kind of the managed application. Possible values are `MarketPlace` and `ServiceCatalog`. Changing this forces a new resource to be created.

Reviewer: ```suggestion
* `kind` - (Required) The kind of the managed application to deply. Possible values are `MarketPlace` and `ServiceCatalog`. Changing this forces a new resource to be created.
```

Author: "to deply" - > "to deploy", right? If yes, I updated

> * `target_resource_group_name` - (Required) The name of the target resource group which includes resources managed by Managed Application.

Reviewer 2: Did you consider naming this `managed_resource_group_name`? I haven't heard this to be referred to as 'target resource group' yet and "managed resource group" would be consistent with `az` as well:

```
$ az managedapp create --help

Command
    az managedapp create : Create a managed application.

Arguments
    --kind              [Required] : The managed application kind. can be marketplace or
                                     servicecatalog.
    --managed-rg-id -m  [Required] : The resource group managed by the managed application.
    --name -n           [Required] : The managed application name.
    --resource-group -g [Required] : The resource group of the managed application.
...
```

Author: @katbyte . I think what seiffert suggests is reasonable.  I found generally it would be named as "managed_resource_group_name" in Azure side after went through the related docs. I think it would be more friendly to azure users. How do you think? If it's fine, I will update it to "managed_resource_group_name"?

Reviewer: I don't think the API's name is good or clear as to what it is (or if it should be created first) but i'll leave it up to you @neil-yechenwei - please make sure the docs are clear what it means either way

the CLi tool is auto generated from swagger and often carries over tragic names & descriptions @seiffert - we often will change/rename properties so they is more obvious as to what they do ie target resource group implies the resource will act upon that group. However in this case its not as egregous as i've seen elsewhere so i'll leave it up to the PR author.

Author: Thanks both. I updated it to "managed_resource_group_name" and updated the document for the description.

> * `managed_resource_group_name` - (Required) The name of the target resource group which holds all the resources that are required by the managed application. Changing this forces a new resource to be created.

Reviewer: I think this should be 
```suggestion
* `managed_resource_group_name` - (Required) The name of the target resource group where all the resources deployed by the managed application will reside. Changing this forces a new resource to be created.
```

Author: updated

> * `application_definition_id` - (Optional) The fully qualified path of managed application definition ID.

Reviewer: ```suggestion
* `application_definition_id` - (Optional) The application definition ID to deply.
```

Author: updated

> * `parameters` - (Optional) The name and value pairs that define the managed application parameters.

Reviewer: ```suggestion
* `parameters` - (Optional) A mapping of name and value pairs to pass to the managed application as parameters.
```

Author: updated

* `plan` - (Optional) One `plan` block as defined below.

* `tags` - (Optional) A mapping of tags to assign to the resource.

---

The `plan` block exports the following:

* `name` - (Required) Specifies the name of the plan from the marketplace.

* `product` - (Required) Specifies the product of the plan from the marketplace.

* `publisher` - (Required) Specifies the publisher of the plan.

* `version` - (Required) Specifies the version of the plan from the marketplace.

* `promotion_code` - (Optional) Specifies the promotion code of the plan. 

## Attributes Reference

The following attributes are exported:

> * `id` - The ID of the Managed Application.

Reviewer 2: I believe it would be convenient to expose the app's outputs as well. Do you think we could add that from the beginning?

Author: According by kt's comments previously, suggests to not add outputs from the beginning. Suggests to add them until user asks. Thus we don't need to maintain the outputs user doesn't need whereas outputs is easily to be added.

Reviewer: What are the app outputs you would like to see @seiffert and in what instance would they be sueful?

Reviewer 2: Sorry for the late response. The app outputs can be anything the app publisher decides to output,  so this would need to be a generic key/value map. 

Outputs could be URLs to a management interface deployed as part of the app or also just configuration values for clients to use the app. E.g. if you would deploy a managed app that contains a database server, the server name could be an output of the app. By making it an attribute of the resource, the server name could be used to provision the databases in that server.

Reviewer 2: Of course, this  functionality can also be developed in a future PR. I was just looking at this for a potential use case of our team where outputs would be good to have.

Reviewer: ahh thanks for explaining that makes a lot of sense, @neil-yechenwei could we get an `managed_app_outputs` map as suggested?

Author: @katbyte , may I ask you mean I need to add a Computed map attribute `managed_app_outputs` which includes `name\resource_group_name\kind\managed_resource_group_name\application_definition_id\parameters\plan\tags` in resource file, right? If yes, could we add a data source file to output these properties?

Author: Added

> * `promotion_code` - (Optional) Specifies the promotion code of the plan. 

Reviewer: ```suggestion
* `promotion_code` - (Optional) Specifies the promotion code to use with the plan. 
```

Author: updated

* `kind` - (Required) The kind of the managed application to deploy. Possible values are `MarketPlace` and `ServiceCatalog`. Changing this forces a new resource to be created.

* `managed_resource_group_name` - (Required) The name of the target resource group where all the resources deployed by the managed application will reside. Changing this forces a new resource to be created.

* `application_definition_id` - (Optional) The application definition ID to deploy.

* `parameters` - (Optional) A mapping of name and value pairs to pass to the managed application as parameters.

* `plan` - (Optional) One `plan` block as defined below.

* `tags` - (Optional) A mapping of tags to assign to the resource.

---

The `plan` block exports the following:

* `name` - (Required) Specifies the name of the plan from the marketplace.

* `product` - (Required) Specifies the product of the plan from the marketplace.

* `publisher` - (Required) Specifies the publisher of the plan.

* `version` - (Required) Specifies the version of the plan from the marketplace.

* `promotion_code` - (Optional) Specifies the promotion code to use with the plan.

> ~> **NOTE:** When `plan` is specified, legal terms must be accepted for this item on this subscription before creating the Managed Application. For example, to accept the terms using PowerShell, please use `Set-AzContext -SubscriptionId "00000000-0000-0000-0000-000000000000"; Get-AzMarketplaceTerms -Publisher "cisco" -Product "meraki-vmx" -Name "meraki-vmx100" | Set-AzMarketplaceTerms -Accept`.

Reviewer: could we make this a bit more generic
```suggestion
~> **NOTE:** When `plan` is specified, legal terms must be accepted for this item on this subscription before creating the Managed Application. For example, to accept the terms using PowerShell, please use `Set-AzContext -SubscriptionId "00000000-0000-0000-0000-000000000000"; Get-AzMarketplaceTerms -Publisher "plan-publisher" -Product "plan-product" -Name "plan-name" | Set-AzMarketplaceTerms -Accept`.
```

Author: updated
